Add `beforeRequest` hook

jstewmon opened the pull request:
This PR build on the leverage of being able to create got instances (thanks @knksmith57!) by adding a `finalize` option, which eases the use of HMAC request signatures, such as AWS V4.

I hope the updated example in the readme makes the motivation clear.

I explored implementing this using an event, but it works better if the hook can be an async function (also demonstrated in the readme).

Along the way, I made a couple of tweaks to other areas of the code, which I included here in separate commits. I can pull those off if they're not wanted, but I thought they were small enough that it would just be easier to propose them alongside this new feature.

szmarczak commented:
**Note:** *creating `got` instances is still in the works.*

Seems good, but I don't see any sense of that. You could just do:

```js
const sign = ...;

const instance = got.create({
	options: got.defaults.options,
	methods: got.defaults.methods,
	handler: (url, options, next) => {
		options.headers['sign'] = sign(options);

		return next(url, options);
	}
});
```

> Along the way, I made a couple of tweaks to other areas of the code, which I included here in separate commits.

Can you do another PR for that? To make everything more clear :)

szmarczak commented on source/request-as-event-emitter.js:
```diff
@@ -228,6 +228,9 @@ module.exports = (options = {}) => {
 
+			if (is.function(options.finalize)) {
+				await options.finalize(options);
+			}
```
If `got.create` wasn't existing, then it'd be very helpful. I think this options is useless.

jstewmon replied:
I'm sorry, but we are completely misunderstanding each other.

If I understand your suggestion regarding a handler, then a custom handler would run before the default handler.

The entire point of this feature is to create an opportunity to add a signature after the request has been normalized, just before it is sent.

Here's a little test that might illustrate the difference:

```js
test('does something useful', async t => {
	const instance = got.create({
		baseUrl: `${s.url}/api/`,
		options: got.defaults.options,
		methods: got.defaults.methods,
		handler: (url, options, next) => {
			t.is(options.hostname, 'example.com')
			t.is(options.path, '/api/?foo=bar')
			return next(url, options);
		}
	});
	await instance('?foo=bar');
});
```
result:
```

  1 test failed

  does something useful

  /Users/jstewmon/dev/forks/got/test/finalize.js:95

   94:     handler: (url, options, next) => {
   95:       t.is(options.hostname, 'example.com')
   96:       t.is(options.path, '/api/?foo=bar')

  Difference:

  - undefined
  + 'example.com'

  Object.handler (test/finalize.js:95:6)
  got (source/create.js:34:20)
  test/finalize.js:100:8
```

szmarczak replied:
Thanks for clarifying this out :)

szmarczak replied:
This should be `.on('beforeRequest', ...)`. Options may be overridden.

jstewmon replied:
I considered doing that in `assignOptions`, but I decided that it was better to leave that choice to the caller, since I think it is pretty easy:

```js
const got = require('got');
const a = got.extend({beforeRequest: opts => { opts.header.foo = 'bar'; }});
const b = a.extend({beforeRequest: opts => { a.defaults.beforeRequest(opts); opts.header.bar = 'foo'; }});
```

If the functions are chained, I don't think there's a convenient way to replace the `beforeRequest` option if that's the behavior one desires.

szmarczak replied:
> we just wrap it in a function that calls both.

@sindresorhus We need to be strict here. Only for `beforeRequest` and `retries`.

> I decided that it was better to leave that choice to the caller, since I think it is pretty easy

It's OK for now. Lots of things is gonna change if #510 gets merged. #510 needs a lot of polishing.

jstewmon replied:
I took a look at #510 and had the same initial reaction as others - I don't get it. :-)

Personally, I don't see providing merge/extend features as creating a lot of leverage. Since the options are accessible on a got instance, the caller can always use a merge strategy of their choosing to create the configuration for a new client. If it were my choice, I'd probably just provide `Object.assign({}, old, new)` semantics for creating new instances.

szmarczak replied:
@jstewmon It's not as easy as it seems to be ;) That's for the future: Got plugins.

sindresorhus commented on readme.md:
```diff
@@ -243,6 +243,15 @@ Determines if a `got.HTTPError` is thrown for error responses (non-2xx status co
 
 If this is disabled, requests that encounter an error status code will be resolved with the `response` instead of throwing. This may be useful if you are checking for resource availability and are expecting error responses.
 
+###### finalize
```
I think `beforeRequest` would be a clearer name.

jstewmon replied:
I'm ok with that. I chose `finalize` to indicate that this is the very last thing that happens before the request is sent.

I'll rename.

jstewmon commented on test/before-request.js:
```diff
@@ -0,0 +1,91 @@
+		{
+			json: true,
+			beforeRequest: async options => {
+				return new Promise(
```
Ah, I didn't see it. Thanks and done!

jstewmon commented on readme.md:
```diff
@@ -243,6 +243,15 @@ Determines if a `got.HTTPError` is thrown for error responses (non-2xx status co
+Type: `Function`<br>
+Default: `undefined`
+
+Called with the normalized request options just before the request is sent. You can modify the object. This is especially useful in conjunction with [`got.extend()`](#instances) and [`got.create()`](advanced-creation.md) when you want to create an API client that uses HMAC-signing.
```
Is it worth noting that changing the body is not recommended, since the content-length has already been set?

sindresorhus replied:
Yes

brandon93s commented:
This is similar in objective to [interceptors](https://github.com/axios/axios#interceptors) in axios. It's also a great place to add support for plugins, or hooks. An API like the following would allow us to expand this to different hooks and provide support for _multiple_ hooks at once:

```js
const plugin = got.hooks.beforeRequest.add(async req => {

});

got.hooks.beforeRequest.remove(plugin);
```

Hooks here could be things like: 

- Authentication 
- Logging / Debugging
	- For example: `chalk()` of network activity in development mode
- Etc

I support the current implementation, but wanted to bring up the alternative above if we want to future proof the implementation.

jstewmon commented:
I actually ended up coming back to got for this after trying to solve the problem with axios interceptors. :-)

axios request interceptors run before their dispatchRequest method, which is where they normalize the request.

But, that's not your point...

I considered allowing the config setting an array or a function, but I decided that it was best to start the conversation around the feature by making it as simple as possible.

If we allow the setting to be an array (without providing a custom interface), I think that might provide better mechanics for anyone wanting to apply a custom merge strategy to the options for a new instance (versus the function wrapping discussed earlier).

If it's conceivable that additional hook points would be added in the future, then we might consider having a top-level `hooks` setting, which is an object where they keys are the hook names and the values are `Array|Function`.

szmarczak commented:
@brandon93s +1

sindresorhus commented:
@szmarczak Events notify while hooks let you transform.

jstewmon commented:
I rebased off master and updated the interface to reflect the discussion about future-proofing the configuration interface for additional hook events.

jstewmon commented on source/index.js:
```diff
@@ -23,6 +23,9 @@ const defaults = {
 			'user-agent': `${pkg.name}/${pkg.version} (https://github.com/sindresorhus/got)`
+		},
+		hooks: {
+			beforeRequest: []
```
I strongly disagree. I'll repeat the rationale I provided earlier in the discussion:
> If we allow the setting to be an array (without providing a custom interface), I think that might provide better mechanics for anyone wanting to apply a custom merge strategy to the options for a new instance (versus the function wrapping discussed earlier).

Requiring the use of a custom interface add no value and prohibits advanced configuration scenarios.

szmarczak replied:
> If we allow the setting to be an array (without providing a custom interface)

OK. But this doesn't change anything.

> Requiring the use of a custom interface adds no value and prohibits advanced configuration scenarios.

I disagree. What advanced configuration scenarios does it prohibit?

> I think that might provide better mechanics for anyone wanting to apply a custom merge strategy to the options for a new instance (versus the function wrapping discussed earlier).

You're right here.


What if defaults were looking like:
```js
{
	options,
	methods,
	handler: (options, hooks, next) => ...
}
```

szmarczak replied:
BTW, do we really need to specify empty hooks in the defaults?

sindresorhus replied:
I prefer just a simple array too.

@szmarczak Making it just an array enables the user to construct it however they want. They can order it based on a condition.

Same reason we have:

```js
got(..., {
	headers: {foo: true}
});
```

Instead of:

```js
got.headers.add({foo: true});
```

szmarczak replied:
ok

szmarczak replied:
> I did it this way, and ensured hooks are defined as arrays during normalization

So it's needed to specify them in the defaults? If not we can just get rid of that.

jstewmon replied:
Ah, I see your point now...

I suspect that there is a performance advantage to defining the defaults.

Without the defaults, an object and `knownHookEvents.length` arrays will be created for every request, resulting in more work for the garbage collector.

Further, I think that dynamically creating the properties will introduce a new [hidden class] for every request. So, I think it is best to keep the default defined.

These are certainly micro optimizations, but I don't see a disadvantage to defining the defaults. Do you?

[hidden class]: https://draft.li/blog/2016/12/22/javascript-engines-hidden-classes/

szmarczak replied:
I don't :) But look: there are other properties that get redefined: [`options.timeout`](https://github.com/sindresorhus/got/blob/98b566444c9dd103af14281ace61fe2bd17d4f49/source/normalize-arguments.js#L179), [`options.followRedirect`](https://github.com/sindresorhus/got/blob/98b566444c9dd103af14281ace61fe2bd17d4f49/source/normalize-arguments.js#L175), [`options.retry`](https://github.com/sindresorhus/got/blob/98b566444c9dd103af14281ace61fe2bd17d4f49/source/normalize-arguments.js#L118) and more. I don't think it's a big deal.

brandon93s replied:
The array approach does come with the caveat that we can't support addition or removal of hooks from an instance since the options are run through an `Object.freeze`. A new instance would have to be created to modify hooks - they **must** be provided at instantiation.

sindresorhus replied:
@brandon93s I would consider that a feature. Immutability good, it prevents a lot of weird bugs.

Do you have any use-case for when you would want to modify hooks after instantiation?

brandon93s replied:
No, nothing in particular comes to mind. I'm comfortable with how it has been implemented here. I prefer the consistent API over additional methods to maintain hooks.

szmarczak commented on source/normalize-arguments.js:
```diff
@@ -8,6 +8,7 @@ const urlToOptions = require('./url-to-options');
 const isFormData = require('./is-form-data');
 
 const RETRY_AFTER_STATUS_CODES = new Set([413, 429, 503]);
+const knownHookEvents = ['beforeRequest'];
```
This should be named `KNOWN_HOOK_EVENTS` to keep the consistency of the code style :)

jstewmon replied:
I saw `RETRY_AFTER_STATUS_CODES` was chosen recently in #508, but I'm wondering if y'all would like to reconsider that choice and make it `retryAfterStatusCodes` as was the alternative suggestion in that discussion.

This is the only place in this package where this naming style is used. I know that uppercase constants is conventional in some languages, but not JS. Further, having two styles for variable names requires a decision to be made about when to use the uppercase style. Is it for all vars declared as `const`? It is for `const` outside of a function?

sindresorhus replied:
@jstewmon I agree. I feel like I never know when to put it in uppercase. Better to just always use normal camelcase, indeed.

sindresorhus commented on readme.md:
```diff
@@ -258,6 +258,24 @@ Determines if a `got.HTTPError` is thrown for error responses (non-2xx status co
 
+###### hooks
+
+Type: `Object<string, Array<Function>`<br>
```
This is missing `>`

sindresorhus commented on readme.md:
```diff
@@ -626,47 +644,31 @@ got('http://unix:/var/run/docker.sock:/containers/json');
+const awsClient = got.extend(
+	{
+		baseUrl: 'https://<api-id>.execute-api.<api-region>.amazonaws.com/<stage>/',
+		beforeRequest: async options => {
```
`beforeRequest` needs to be nested in `hooks`

sindresorhus commented on readme.md:
```diff
@@ -626,47 +644,31 @@ got('http://unix:/var/run/docker.sock:/containers/json');
 
-request(`https://${config.host}/production/`, {
-	// All usual `got` options
+const response = await awsClient('/endpoint/path', {
```
@szmarczak I don't think we should add variables that are not defined in the example. Ideally, examples should be copy-paste runnnable. I also think the comment helps make it clearer.

szmarczak replied:
Okay then :)

sindresorhus commented on source/request-as-event-emitter.js:
```diff
@@ -241,6 +241,11 @@ module.exports = (options = {}) => {
 			}
 
+			for (const hook of options.hooks.beforeRequest) {
+				// eslint-disable-next-line no-await-in-loop
```
@szmarczak Let's try not to nitpick too much. The super minor stuff we can just fix ourselves when merging.

szmarczak replied:
I just want to note that we should keep the style of coding. That's all.

jstewmon commented on readme.md:
```diff
@@ -258,6 +258,24 @@ Determines if a `got.HTTPError` is thrown for error responses (non-2xx status co
+
+See the [AWS section](#aws) for an example.
+
+**Note**: Modifying the `body` is not recommended because the `content-length` header has already been computed and assigned.
```
There are a few ways this could be strictly prohibited, but I think this is case where it's best to provide guidance but not prohibit advanced usage. Someone may have a valid reason for modifying the body, so this note is here to hint they they probably need to reassign `content-length` if they change body.

szmarczak replied:
I agree. Or maybe just move

https://github.com/jstewmon/got/blob/c0aa4d74b9d9d94c858520586b5def68440cd9a9/source/request-as-event-emitter.js#L246

right after `try` before it calculates the body size?

jstewmon replied:
That would violate the expectation that got will make no further changes to the options before the request is sent. ;-)

szmarczak replied:
> That would violate the expectation that got will make no further changes to the options before the request is sent.

But that's only one header... I think that'd be fine :)

@sindresorhus What do you think?

jstewmon replied:
If that is a signed header, it will break the whole thing. It's a big deal.

jstewmon replied:
I updated to say that Got will make no further changes to the request before it is sent.

I used the signing feature as an example of why I did not want to move the call before the last attempt to set `content-length`, but we shouldn't specifically mention request signing in the docs for `beforeRequest`, since it's not tightly coupled to request signing.

szmarczak commented on readme.md:
```diff
@@ -270,7 +270,7 @@ Hooks allow modifications during the request lifecycle. Hook functions may be as
 Default: `[]`
 
-Called with the normalized request options just before the request is sent. You can modify the object. This is especially useful in conjunction with [`got.extend()`](#instances) and [`got.create()`](advanced-creation.md) when you want to create an API client that uses HMAC-signing.
+Called with the normalized request options just before the request is sent. Got will make no further changes to the request before it is sent. This is especially useful in conjunction with [`got.extend()`](#instances) and [`got.create()`](advanced-creation.md) when you want to create an API client that uses HMAC-signing.
```
Butter butter. Don't repeat yourself ;)

> just before the request is sent

it says that for itself :)

sindresorhus commented:
This looks great now. Thank you so much for working on this, @jstewmon :)

sindresorhus commented:
Note discussion in https://github.com/sindresorhus/got/pull/516#discussion_r202513672
